text-serializer-legacy/src/main/java/net/kyori/adventure/text/serializer/legacy/LegacyCodeConverter.java

```diff
@@ -0,0 +1,171 @@
+      return new DecodedFormat(foundFormat, getFormats().get(getAllLegacyChars().indexOf(legacy)));
+    } else if (foundFormat == FormatCodeType.BUNGEECORD_UNUSUAL_HEX) {
+      final StringBuilder foundHex = new StringBuilder(6);
+      for (int i = pos - 1; i >= pos - 11; i -= 2) {
+        foundHex.append(input.charAt(i));
+      }
+      final @Nullable TextColor parsed = tryParseHexColor(foundHex.reverse().toString());
+      if (parsed != null) {
+        return new DecodedFormat(foundFormat, parsed);
+      }
+    }
+    return null;
+  }
+
+  private static @Nullable TextColor tryParseHexColor(final String hexDigits) {
+    try {
+      final int color = Integer.parseInt(hexDigits, 16);
+      return TextColor.color(color);
+    } catch (final NumberFormatException ex) {
+      return null;
+    }
+  }
+
+  private static boolean isHexTextColor(final TextFormat format) {
+    return format instanceof TextColor && !(format instanceof NamedTextColor);
+  }
+
+  String toLegacyCode(TextFormat format) {
+    if (isHexTextColor(format)) {
+      final TextColor color = (TextColor) format;
+      if (this.hexColours) {
+        final String hex = String.format("%06x", color.value());
+        if (this.useTerriblyStupidHexFormat) {
+          // ah yes, wonderful. A 14 digit long completely unreadable string.
+          final StringBuilder legacy = new StringBuilder(String.valueOf(LEGACY_BUNGEE_HEX_CHAR));
+          for (int i = 0, length = hex.length(); i < length; i++) {
+            legacy.append(this.character).append(hex.charAt(i));
+          }
+          return legacy.toString();
+        } else {
+          // this is a bit nicer, hey?
+          return this.hexCharacter + hex;
+        }
+      } else {
+        // if we are not using hex colours, then convert the hex colour
+        // to the "nearest" possible named/standard text colour
+        format = nearestColorTo(color);
+      }
+    }
+    final int index = getFormats().indexOf(format);
+    return Character.toString(getAllLegacyChars().charAt(index));
+  }
+
+  /**
+   * Find the named colour nearest to the provided colour.
+   *
+   * @param any colour to match
+   * @return nearest named colour. will always return a value
+   * @since 4.14.0
+   */
+  public @NotNull TextColor nearestColorTo(final @NotNull TextColor any) {
+    if (any instanceof NamedTextColor) {
+      return any;
+    }
+
+    requireNonNull(any, "color");
+
+    float matchedDistance = Float.MAX_VALUE;
+    final List<? extends TextColor> values = getLegacyColors();
+    TextColor match = values.get(0);
+    for (int i = 0, length = values.size(); i < length; i++) {
+      final TextColor potential = values.get(i);
+      final float distance = distance(any.asHSV(), potential.asHSV());
+      if (distance < matchedDistance) {
+        match = potential;
+        matchedDistance = distance;
+      }
+      if (distance == 0) {
+        break; // same colour! whoo!
+      }
+    }
+    return match;
```

Comment:
Instead of duplicating this logic, we should refactor the original method in `NamedTextColor` to be something like `public <C extends TextColor> @NotNull C nearestColorTo(final @NotNull List<C> values, final @NotNull TextColor any)` - have the existing method in that class call it with `nearestColorTo(VALUES, any)` and have this new class call it as well.